File: sdks/python/hooksniff/resources/endpoint.py
"""
HookSniff SDK — Endpoint Resource

Adapted from Svix SDK architecture.
"""

import typing as t
from dataclasses import dataclass

from .. import models
from ..models import (
    EndpointIn,
    EndpointOut,
    EndpointUpdate,
    EndpointSecretOut,
    EndpointStats,
    EndpointHealth,
    ListResponseEndpointOut,
)
from .common import ApiBase, BaseOptions, serialize_params


@dataclass
class EndpointListOptions(BaseOptions):
    limit: t.Optional[int] = None
    """Limit the number of returned items"""
    iterator: t.Optional[str] = None
    """The iterator returned from a prior invocation"""
    order: t.Optional[models.Ordering] = None
    """The sorting order of the returned items"""
    is_active: t.Optional[bool] = None
    """Filter by active status"""

    def _query_params(self) -> t.Dict[str, str]:
        return serialize_params(
            {
                "limit": self.limit,
                "iterator": self.iterator,
                "order": self.order,
                "is_active": self.is_active,
            }
        )


@dataclass
class EndpointCreateOptions(BaseOptions):
    idempotency_key: t.Optional[str] = None

    def _header_params(self) -> t.Dict[str, str]:
        return serialize_params(
            {
                "idempotency-key": self.idempotency_key,
            }
        )


class EndpointAsync(ApiBase):
    async def list(
        self, options: EndpointListOptions = EndpointListOptions()
    ) -> ListResponseEndpointOut:
        """List all endpoints for the authenticated user."""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/endpoints",
            query_params=options._query_params(),
        )
        return ListResponseEndpointOut(**response.json())

    async def create(
        self, endpoint_in: EndpointIn, options: EndpointCreateOptions = EndpointCreateOptions()
    ) -> EndpointOut:
        """Create a new endpoint."""
        response = await self._request_asyncio(
            method="post",
            path="/api/v1/endpoints",
            header_params=options._header_params(),
            json_body=endpoint_in.to_dict() if hasattr(endpoint_in, "to_dict") else endpoint_in.__dict__,
        )
        return EndpointOut(**response.json())

    async def get(self, endpoint_id: str) -> EndpointOut:
        """Get an endpoint by ID."""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/endpoints/{endpoint_id}",
            path_params={"endpoint_id": endpoint_id},
        )
        return EndpointOut(**response.json())

    async def update(self, endpoint_id: str, endpoint_update: EndpointUpdate) -> EndpointOut:
        """Update an endpoint."""
        response = await self._request_asyncio(
            method="put",
            path="/api/v1/endpoints/{endpoint_id}",
            path_params={"endpoint_id": endpoint_id},
            json_body=endpoint_update.__dict__,
        )
        return EndpointOut(**response.json())

    async def delete(self, endpoint_id: str) -> None:
        """Delete an endpoint."""
        await self._request_asyncio(
            method="delete",
            path="/api/v1/endpoints/{endpoint_id}",
            path_params={"endpoint_id": endpoint_id},
        )

    async def get_secret(self, endpoint_id: str) -> EndpointSecretOut:
        """Get the signing secret for an endpoint."""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/endpoints/{endpoint_id}/secret",
            path_params={"endpoint_id": endpoint_id},
        )
        return EndpointSecretOut(**response.json())

    async def rotate_secret(
        self, endpoint_id: str, options: EndpointCreateOptions = EndpointCreateOptions()
    ) -> EndpointSecretOut:
        """Rotate the signing secret for an endpoint."""
        response = await self._request_asyncio(
            method="post",
            path="/api/v1/endpoints/{endpoint_id}/rotate-secret",
            path_params={"endpoint_id": endpoint_id},
            header_params=options._header_params(),
        )
        return EndpointSecretOut(**response.json())

    async def get_stats(self, endpoint_id: str) -> EndpointStats:
        """Get delivery stats for an endpoint."""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/endpoints/{endpoint_id}/stats",
            path_params={"endpoint_id": endpoint_id},
        )
        return EndpointStats(**response.json())

    async def get_health(self, endpoint_id: str) -> EndpointHealth:
        """Get health info for an endpoint."""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/endpoints/{endpoint_id}/health",
            path_params={"endpoint_id": endpoint_id},
        )
        return EndpointHealth(**response.json())


class Endpoint(ApiBase):
    def list(
        self, options: EndpointListOptions = EndpointListOptions()
    ) -> ListResponseEndpointOut:
        """List all endpoints for the authenticated user."""
        response = self._request_sync(
            method="get",
            path="/api/v1/endpoints",
            query_params=options._query_params(),
        )
        return ListResponseEndpointOut(**response.json())

    def create(
        self, endpoint_in: EndpointIn, options: EndpointCreateOptions = EndpointCreateOptions()
    ) -> EndpointOut:
        """Create a new endpoint."""
        response = self._request_sync(
            method="post",
            path="/api/v1/endpoints",
            header_params=options._header_params(),
            json_body=endpoint_in.to_dict() if hasattr(endpoint_in, "to_dict") else endpoint_in.__dict__,
        )
        return EndpointOut(**response.json())

    def get(self, endpoint_id: str) -> EndpointOut:
        """Get an endpoint by ID."""
        response = self._request_sync(
            method="get",
            path="/api/v1/endpoints/{endpoint_id}",
            path_params={"endpoint_id": endpoint_id},
        )
        return EndpointOut(**response.json())

    def update(self, endpoint_id: str, endpoint_update: EndpointUpdate) -> EndpointOut:
        """Update an endpoint."""
        response = self._request_sync(
            method="put",
            path="/api/v1/endpoints/{endpoint_id}",
            path_params={"endpoint_id": endpoint_id},
            json_body=endpoint_update.__dict__,
        )
        return EndpointOut(**response.json())

    def delete(self, endpoint_id: str) -> None:
        """Delete an endpoint."""
        self._request_sync(
            method="delete",
            path="/api/v1/endpoints/{endpoint_id}",
            path_params={"endpoint_id": endpoint_id},
        )

    def get_secret(self, endpoint_id: str) -> EndpointSecretOut:
        """Get the signing secret for an endpoint."""
        response = self._request_sync(
            method="get",
            path="/api/v1/endpoints/{endpoint_id}/secret",
            path_params={"endpoint_id": endpoint_id},
        )
        return EndpointSecretOut(**response.json())

    def rotate_secret(
        self, endpoint_id: str, options: EndpointCreateOptions = EndpointCreateOptions()
    ) -> EndpointSecretOut:
        """Rotate the signing secret for an endpoint."""
        response = self._request_sync(
            method="post",
            path="/api/v1/endpoints/{endpoint_id}/rotate-secret",
            path_params={"endpoint_id": endpoint_id},
            header_params=options._header_params(),
        )
        return EndpointSecretOut(**response.json())

    def get_stats(self, endpoint_id: str) -> EndpointStats:
        """Get delivery stats for an endpoint."""
        response = self._request_sync(
            method="get",
            path="/api/v1/endpoints/{endpoint_id}/stats",
            path_params={"endpoint_id": endpoint_id},
        )
        return EndpointStats(**response.json())

    def get_health(self, endpoint_id: str) -> EndpointHealth:
        """Get health info for an endpoint."""
        response = self._request_sync(
            method="get",
            path="/api/v1/endpoints/{endpoint_id}/health",
            path_params={"endpoint_id": endpoint_id},
        )
        return EndpointHealth(**response.json())

File: sdks/kotlin/lib/src/main/kotlin/Endpoint.kt
// this file is @generated
package com.hooksniff.kotlin

import com.hooksniff.kotlin.models.EndpointHeadersIn
import com.hooksniff.kotlin.models.EndpointHeadersOut
import com.hooksniff.kotlin.models.EndpointHeadersPatchIn
import com.hooksniff.kotlin.models.EndpointIn
import com.hooksniff.kotlin.models.EndpointOut
import com.hooksniff.kotlin.models.EndpointPatch
import com.hooksniff.kotlin.models.EndpointSecretOut
import com.hooksniff.kotlin.models.EndpointSecretRotateIn
import com.hooksniff.kotlin.models.EndpointUpdate
import com.hooksniff.kotlin.models.EventExampleIn
import com.hooksniff.kotlin.models.ListResponseEndpointOut
import com.hooksniff.kotlin.models.MessageOut
import com.hooksniff.kotlin.models.Ordering
import okhttp3.Headers

data class EndpointListOptions(
    /** Limit the number of returned items */
    val limit: ULong? = null,
    /** The iterator returned from a prior invocation */
    val iterator: String? = null,
    /** The sorting order of the returned items */
    val order: Ordering? = null,
)

data class EndpointCreateOptions(val idempotencyKey: String? = null)

data class EndpointRotateSecretOptions(val idempotencyKey: String? = null)

data class EndpointSendExampleOptions(val idempotencyKey: String? = null)

class Endpoint(private val client: HookSniffHttpClient) {
    /** List the application's endpoints. */
    suspend fun list(
        appId: String,
        options: EndpointListOptions = EndpointListOptions(),
    ): ListResponseEndpointOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/app/$appId/endpoint")
        options.limit?.let { url.addQueryParameter("limit", serializeQueryParam(it)) }
        options.iterator?.let { url.addQueryParameter("iterator", it) }
        options.order?.let { url.addQueryParameter("order", serializeQueryParam(it)) }
        return client.executeRequest<Any, ListResponseEndpointOut>("GET", url.build())
    }

    /**
     * Create a new endpoint for the application.
     *
     * When `secret` is `null` the secret is automatically generated (recommended).
     */
    suspend fun create(
        appId: String,
        endpointIn: EndpointIn,
        options: EndpointCreateOptions = EndpointCreateOptions(),
    ): EndpointOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/app/$appId/endpoint")
        val headers = Headers.Builder()
        options.idempotencyKey?.let { headers.add("idempotency-key", it) }

        return client.executeRequest<EndpointIn, EndpointOut>(
            "POST",
            url.build(),
            headers = headers.build(),
            reqBody = endpointIn,
        )
    }

    /** Get an endpoint. */
    suspend fun get(appId: String, endpointId: String): EndpointOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/app/$appId/endpoint/$endpointId")
        return client.executeRequest<Any, EndpointOut>("GET", url.build())
    }

    /** Update an endpoint. */
    suspend fun update(
        appId: String,
        endpointId: String,
        endpointUpdate: EndpointUpdate,
    ): EndpointOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/app/$appId/endpoint/$endpointId")

        return client.executeRequest<EndpointUpdate, EndpointOut>(
            "PUT",
            url.build(),
            reqBody = endpointUpdate,
        )
    }

    /** Delete an endpoint. */
    suspend fun delete(appId: String, endpointId: String) {
        val url = client.newUrlBuilder().encodedPath("/api/v1/app/$appId/endpoint/$endpointId")
        client.executeRequest<Any, Boolean>("DELETE", url.build())
    }

    /** Partially update an endpoint. */
    suspend fun patch(
        appId: String,
        endpointId: String,
        endpointPatch: EndpointPatch,
    ): EndpointOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/app/$appId/endpoint/$endpointId")

        return client.executeRequest<EndpointPatch, EndpointOut>(
            "PATCH",
            url.build(),
            reqBody = endpointPatch,
        )
    }

    /** Get the additional headers to be sent with the webhook. */
    suspend fun getHeaders(appId: String, endpointId: String): EndpointHeadersOut {
        val url =
            client.newUrlBuilder().encodedPath("/api/v1/app/$appId/endpoint/$endpointId/headers")
        return client.executeRequest<Any, EndpointHeadersOut>("GET", url.build())
    }

    /** Set the additional headers to be sent with the webhook. */
    suspend fun updateHeaders(
        appId: String,
        endpointId: String,
        endpointHeadersIn: EndpointHeadersIn,
    ) {
        val url =
            client.newUrlBuilder().encodedPath("/api/v1/app/$appId/endpoint/$endpointId/headers")

        client.executeRequest<EndpointHeadersIn, Boolean>(
            "PUT",
            url.build(),
            reqBody = endpointHeadersIn,
        )
    }

    /** Partially set the additional headers to be sent with the webhook. */
    suspend fun patchHeaders(
        appId: String,
        endpointId: String,
        endpointHeadersPatchIn: EndpointHeadersPatchIn,
    ) {
        val url =
            client.newUrlBuilder().encodedPath("/api/v1/app/$appId/endpoint/$endpointId/headers")

        client.executeRequest<EndpointHeadersPatchIn, Boolean>(
            "PATCH",
            url.build(),
            reqBody = endpointHeadersPatchIn,
        )
    }

    /**
     * Get the endpoint's signing secret.
     *
     * This is used to verify the authenticity of the webhook. For more information please refer to
     * [the consuming webhooks docs](https://docs.hooksniff.com/consuming-webhooks/).
     */
    suspend fun getSecret(appId: String, endpointId: String): EndpointSecretOut {
        val url =
            client.newUrlBuilder().encodedPath("/api/v1/app/$appId/endpoint/$endpointId/secret")
        return client.executeRequest<Any, EndpointSecretOut>("GET", url.build())
    }

    /**
     * Rotates the endpoint's signing secret.
     *
     * The previous secret will remain valid for the next 24 hours.
     */
    suspend fun rotateSecret(
        appId: String,
        endpointId: String,
        endpointSecretRotateIn: EndpointSecretRotateIn,
        options: EndpointRotateSecretOptions = EndpointRotateSecretOptions(),
    ) {
        val url =
            client
                .newUrlBuilder()
                .encodedPath("/api/v1/app/$appId/endpoint/$endpointId/secret/rotate")
        val headers = Headers.Builder()
        options.idempotencyKey?.let { headers.add("idempotency-key", it) }

        client.executeRequest<EndpointSecretRotateIn, Boolean>(
            "POST",
            url.build(),
            headers = headers.build(),
            reqBody = endpointSecretRotateIn,
        )
    }

    /** Send an example message for an event. */
    suspend fun sendExample(
        appId: String,
        endpointId: String,
        eventExampleIn: EventExampleIn,
        options: EndpointSendExampleOptions = EndpointSendExampleOptions(),
    ): MessageOut {
        val url =
            client
                .newUrlBuilder()
                .encodedPath("/api/v1/app/$appId/endpoint/$endpointId/send-example")
        val headers = Headers.Builder()
        options.idempotencyKey?.let { headers.add("idempotency-key", it) }

        return client.executeRequest<EventExampleIn, MessageOut>(
            "POST",
            url.build(),
            headers = headers.build(),
            reqBody = eventExampleIn,
        )
    }
}
